kotlin/internal/jvm/compile.bzl

```diff
@@ -237,81 +267,174 @@ def kt_jvm_compile_action(ctx, rule_kind, output_jar, compile_jar):
+           exports = [d[JavaInfo] for d in getattr(ctx.attr, "exports", [])],
+           neverlink = getattr(ctx.attr, "neverlink", False),
+        )
+        java_infos.append(kt_java_info)
+
+    # Build Java
+    # If there is Java source or KAPT generated Java source compile that Java and fold it into
+    # the final ABI jar. Otherwise just use the KT ABI jar as final ABI jar.
+    if srcs.java or generated_src_jars:
+        toolchain = ctx.toolchains[_TOOLCHAIN_TYPE]
+        javac_opts = [
+            # Disable option linting, it will complain about the source.
+            "-Xlint:-options",
+            "-source " + toolchain.jvm_target,
+            "-target " + toolchain.jvm_target
+        ]
+
+        # Kotlin takes care of annotation processing.
+        if len(srcs.kt) > 0:
+            javac_opts += ["-proc:none"]
+
+        java_info = java_common.compile(
+            ctx,
+            source_files = srcs.java,
+            source_jars = generated_src_jars,
+            output = ctx.actions.declare_file(ctx.label.name + "-java.jar"),
+            javac_opts = javac_opts,
+            deps = compile_deps.deps + kt_stubs_for_java,
+            java_toolchain = toolchains.java,
+            host_javabase = toolchains.java_runtime,
+            neverlink = getattr(ctx.attr, "neverlink", False)
+        )
+        compile_jars = compile_jars + [
+            jars.ijar
+            for jars in java_info.outputs.jars
+        ]
+        output_jars = output_jars + [
+            jars.class_jar
+            for jars in java_info.outputs.jars
+        ]
+        java_infos.append(java_info)
+
+    compile_jar = ctx.actions.declare_file(ctx.label.name + ".abi.jar")
+    output_jar = ctx.actions.declare_file(ctx.label.name + ".jar")
+
+    # Merge ABI jars into final compile jar.
+    _fold_jars_action(
+        ctx,
         rule_kind = rule_kind,
-        toolchains = toolchains,
-        dirs = dirs,
-        srcs = srcs,
-        friend = friend,
-        compile_deps = compile_deps,
-        annotation_processors = annotation_processors,
-        transitive_runtime_jars = transitive_runtime_jars,
-        plugins = plugins,
-        outputs = {
-            "output": output_jar,
-            "kotlin_output_jdeps": ctx.outputs.jdeps,
-            "kotlin_output_srcjar": ctx.outputs.srcjar,
-        },
+        output_jar = compile_jar,
+        action_type = "Abi",
+        input_jars = compile_jars,
+    )
+
+    output_jars.extend(kt_jvm_produce_output_jar_actions(ctx, rule_kind))
+    # Merge outputs into final runtime jar.
+    _fold_jars_action(
+        ctx,
+        rule_kind = rule_kind,
+        output_jar = output_jar,
+        action_type = "Runtime",
+        input_jars = output_jars,
     )
 
+    mergee_jdeps = []
+    for java_info in java_infos:
+        if java_info.outputs.jdeps:
+            mergee_jdeps.append(java_info.outputs.jdeps);
+
+#    jdeps = _run_merge_jdeps_action(
```

Comment:
Need to figure out what to do about jdeps.

It seems pretty slow, on our fork I've just removed it temporarily, but I see two possibilities
1. Break it out of the Kotlin Builder and extract it to its own Builder and run it on the output jar. We should probably run it on the compile jar also so we can support classpath shortening.
1. Implement a Kotlin compiler plugin to generate the Kotlin half of the jdeps and then write a new builder to merge the two jdeps (I already have the jdeps merger in our fork). Ideally `java_common.merge()` or a new API would support this use case. I can also see about contributing something to Bazel for this.

```
Task timings for //src/test/data/jvm/kapt:ap_kotlin_mixed (total: 1322 ms):
|   * expand sources: 0 ms
|   * kapt (com.google.auto.value.processor.AutoValueProcessor): 546 ms
|   * compile classes: 764 ms
|     * kotlinc: 241 ms
|     * javac: 83 ms
|     * create jar: 45 ms
|     * produce src jar: 6 ms
|     * generate jdeps: 389 ms
```

Reply:
I think it's important to note that jdeps never actually worked properly for Kotlin. It does't handle things like inline methods for example.

Reply:
It is better than nothing, unfortunately. We need a kdeps so that we can get strict deps working.